pkg/storage/tsdb/ref_cache_test.go

```diff
@@ -0,0 +1,113 @@
+	ls2 := []labels.Label{{Name: "a", Value: "2"}}
+
+	c := NewRefCache()
+	_, ok := c.GetRef(0, ls1)
+	assert.Equal(t, false, ok)
+
+	_, ok = c.GetRef(0, ls2)
+	assert.Equal(t, false, ok)
+
+	c.SetRef(0, ls1, 1)
+	ref, ok := c.GetRef(0, ls1)
+	assert.Equal(t, true, ok)
+	assert.Equal(t, uint64(1), ref)
+
+	_, ok = c.GetRef(0, ls2)
+	assert.Equal(t, false, ok)
+
+	c.SetRef(0, ls2, 2)
+	ref, ok = c.GetRef(0, ls2)
+	assert.Equal(t, true, ok)
+	assert.Equal(t, uint64(2), ref)
+
+	// Overwrite a value with a new one
+	c.SetRef(0, ls2, 3)
+	ref, ok = c.GetRef(0, ls2)
+	assert.Equal(t, true, ok)
+	assert.Equal(t, uint64(3), ref)
+}
+
+func TestRefCache_ShouldCorrectlyHandleFingerprintCollisions(t *testing.T) {
+	// The two following series have the same FastFingerprint=e002a3a451262627
+	ls1 := []labels.Label{{Name: labels.MetricName, Value: "fast_fingerprint_collision"}, {Name: "app", Value: "l"}, {Name: "uniq0", Value: "0"}, {Name: "uniq1", Value: "1"}}
+	ls2 := []labels.Label{{Name: labels.MetricName, Value: "fast_fingerprint_collision"}, {Name: "app", Value: "m"}, {Name: "uniq0", Value: "1"}, {Name: "uniq1", Value: "1"}}
+
+	c := NewRefCache()
+	c.SetRef(0, ls1, 1)
+	c.SetRef(0, ls2, 2)
+
+	ref, ok := c.GetRef(0, ls1)
+	assert.Equal(t, true, ok)
+	assert.Equal(t, uint64(1), ref)
+
+	ref, ok = c.GetRef(0, ls2)
+	assert.Equal(t, true, ok)
+	assert.Equal(t, uint64(2), ref)
+}
+
+func TestRefCache_Purge(t *testing.T) {
+	series := [][]labels.Label{
+		{{Name: "a", Value: "1"}},
+		{{Name: "a", Value: "2"}},
+		// The two following series have the same FastFingerprint=e002a3a451262627
+		{{Name: labels.MetricName, Value: "fast_fingerprint_collision"}, {Name: "app", Value: "l"}, {Name: "uniq0", Value: "0"}, {Name: "uniq1", Value: "1"}},
+		{{Name: labels.MetricName, Value: "fast_fingerprint_collision"}, {Name: "app", Value: "m"}, {Name: "uniq0", Value: "1"}, {Name: "uniq1", Value: "1"}},
+	}
+
+	// Run the same test for increasing TTL values
+	for ttl := 0; ttl <= len(series); ttl++ {
+		now := int64(100)
+		c := NewRefCache()
+
+		// Set the series to the cache with decreasing timestamps
+		for i := 0; i < len(series); i++ {
+			c.SetRef(now-int64(i), series[i], uint64(i))
+		}
+
+		c.Purge(now, int64(ttl))
+
+		// Check retained and purged entries
+		for i := 0; i <= ttl && i < len(series); i++ {
+			ref, ok := c.GetRef(now, series[i])
+			assert.Equal(t, true, ok)
+			assert.Equal(t, uint64(i), ref)
+		}
+
+		for i := ttl + 1; i < len(series); i++ {
+			_, ok := c.GetRef(now, series[i])
+			assert.Equal(t, false, ok)
+		}
+	}
+}
+
+func BenchmarkRefCache_SetRef(b *testing.B) {
+	ls := []labels.Label{{Name: "a", Value: "1"}}
+	c := NewRefCache()
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		c.SetRef(0, ls, uint64(i))
+	}
+}
+
+func BenchmarkRefCache_GetRef(b *testing.B) {
+	ls := []labels.Label{{Name: "a", Value: "1"}}
+	c := NewRefCache()
+	c.SetRef(0, ls, 1)
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		c.GetRef(0, ls)
+	}
+}
```

Comment:
Given you are using only 1 series, these benchmarks doesn't seem to be pushing it hard enough.